src/cascade/input_data/configuration/builder.py

```diff
@@ -26,6 +58,218 @@ def initial_context_from_epiviz(configuration):
+            elif mul_cov_config.mulcov_type == "meas_std":
+                add_to_integrand.std_covariate_multipliers.append(covariate_multiplier)
+            else:
+                raise RuntimeError(f"mulcov_type isn't among the three {configuration.mulcov_type}")
+
+
+def reference_value_for_covariate_mean_all_values(cov_df):
+    """
+    Strategy for choosing reference value for country covariate.
+    This one takes the mean of all incoming covariate values.
+    """
+    return float(cov_df["mean_value"].mean())
+
+
+def covariate_to_measurements_nearest_favoring_same_year(measurements, covariates):
+    """
+    Given a covariate that might not cover all of the age and time range
+    of the measurements select a covariate value for each measurement.
+    This version chooses the covariate value whose mean age and time
+    is closest to the mean age and time of the measurement in the same
+    year. If that isn't found, it picks the covariate that is closest
+    in age and time in the nearest year. In the case of a tie for distance,
+    it averages.
+
+    Args:
+        measurements (pd.DataFrame):
+            Columns include ``age_lower``, ``age_upper``, ``time_lower``,
+            ``time_upper``. All others are ignored.
+        covariates (pd.DataFrame):
+            Columns include ``age_lower``, ``age_upper``, ``time_lower``,
+            ``time_upper``, and ``value``.
+
+    Returns:
+        pd.Series: One row for every row in the measurements.
+    """
+    # Rescaling the age by 120 means that the nearest age within the year
+    # will always be closer than the nearest time across a full year.
+    tree = spatial.KDTree(list(zip(
+        covariates[["age_lower", "age_upper"]].mean(axis=1) / 240,
+        covariates[["time_lower", "time_upper"]].mean(axis=1),
+        covariates["x_sex"]
+    )))
+    _, indices = tree.query(list(zip(
+        measurements[["age_lower", "age_upper"]].mean(axis=1) / 240,
+        measurements[["time_lower", "time_upper"]].mean(axis=1),
+        measurements["x_sex"],
+    )))
+    return pd.Series(covariates.iloc[indices]["mean_value"].values, index=measurements.index)
+
+
+def convert_age_year_ids_to_ranges(with_ids_df, age_groups_df):
+    """
+    Converts ``age_group_id`` into ``age_lower`` and ``age_upper`` and
+    ``year_id`` into ``time_lower`` and ``time_upper``. This treats the year
+    as a range from start of year to start of the next year.
+    Also converts sex_id=[1, 2, 3] into x_sex=[-0.5, 0.5, 0].
+
+    Args:
+        with_ids_df (pd.DataFrame): Has ``age_group_id`` and ``year_id``.
+        age_groups_df (pd.DataFrame): Has columns ``age_group_id``,
+            ``age_group_years_start``, and ``age_group_years_end``.
+
+    Returns:
+        pd.DataFrame: New pd.DataFrame with four added columns and in the same
+            order as the input dataset.
+    """
+    # Is this mapping right? Is -0.5 men or women?
```

Comment:
-0.5 is women; that is, I remember Alex saying that, and here is an excerpt from Alex's internal doc (https://hub.ihme.washington.edu/display/DataScience/Internal+Docs).  I don't know if it's doc'd on Brad's web.

"For male runs, reference=0.5 and max_difference=0.6. For female runs, reference=-0.5 and max_difference=0.6. These settings are very important because they are the only thing that differentiates the male from female runs."